rq/queue.py

```diff
@@ -635,7 +638,9 @@ def enqueue_dependents(self, job, pipeline=None):
                     )
                 ]
 
-                pipe.multi()
+                # only invoke .multi() if all dependent jobs' allow_failure are False-y
+                if not all([job.allow_failure for job in jobs_to_enqueue]):
+                    pipe.multi()
```

Comment:
Sorry, I'm not sure I understand why this is needed.

Reply:
Calling `pipe.multi()` here when any of the `jobs_to_enqueue` has `allow_failure=True` will cause `test_worker.py::TestWorker::test_job_dependency` to fail.

Reply:
I think this may be caused by the fact that there are [other commands](https://github.com/rq/rq/pull/1630/files#diff-50984f276e8a6fea5f4ba4938afb6f4fbc544fd3751fbd16027dc0a326d6405eR987) in the pipeline that are not yet executed.

I think we can simplify this by ensuring that `enqueue_dependents` are always called in a clean pipeline. Can you call `pipeline.execute()` prior to calling `enqueue_dependents`? This will make the logic easier to follow and troubleshoot.

Reply:
Sure thing, will make this change and see how it does Monday.

Reply:
@selwin I attempted this change (added a `pipeline.execute()` immediately before `queue.enqueue_dependents(job, pipeline=pipeline)` and removed the conditional statement), but it didn't not behave as expected (job still ends up as deferred). So far, the only solution was to conditionally execute the `pipe.multi()` command. Thoughts?

Reply:
Did you put `pipeline.execute()` before or after `pipeline.watch()`? I think it has to be called before `watch()`. https://github.com/redis/redis-py#pipelines

Reply:
The other alternative here is to call `enqueue_dependents` without `pipeline` so it can call `watch()` on its own (since there will always be two round trips to Redis anyway).

Reply:
> Did you put `pipeline.execute()` before or after `pipeline.watch()`? I think it has to be called before `watch()`. https://github.com/redis/redis-py#pipelines
> The other alternative here is to call enqueue_dependents without pipeline so it can call watch() on its own (since there will always be two round trips to Redis anyway).

I've tried both ways (with removing the conditional statement around the `pipe.multi()` call in `enqueue_dependents`):
```python3
with suppress(redis.exceptions.ConnectionError):
    pipeline.execute()
# if dependencies are inserted after enqueue_dependents
# a WatchError is thrown by execute()
pipeline.watch(job.dependents_key)
# enqueue_dependents calls multi() on the pipeline!
queue.enqueue_dependents(job, pipeline=pipeline)
```
And
```python3
with suppress(redis.exceptions.ConnectionError):
    pipeline.execute()
# if dependencies are inserted after enqueue_dependents
# a WatchError is thrown by execute()
pipeline.watch(job.dependents_key)
# enqueue_dependents calls multi() on the pipeline!
queue.enqueue_dependents(job)
```

I think I'm going to have to call it quits on my efforts soon because I've already spent too much time trying to fix the dependency issues with this package. I'm sadly probably going to have to move on to another package.

Reply:
@caffeinatedMike no worries, thanks for your efforts!

Reply:
I'll try to look into this when I can, am super busy now.